Arm/Memory/SeparateAutomation.lean

```diff
@@ -0,0 +1,192 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Siddharth Bhat
+
+In this file, we define proof automation for separation conditions of memory.
+
+References:
+- https://github.com/leanprover/lean4/blob/240ebff549a2cf557f9abe9568f5de885f13e50d/src/Lean/Elab/Tactic/Omega/OmegaM.lean
+- https://github.com/leanprover/lean4/blob/240ebff549a2cf557f9abe9568f5de885f13e50d/src/Lean/Elab/Tactic/Omega/Frontend.lean
+-/
+import Arm
+import Arm.Memory.MemoryProofs
+import Arm.BitVec
+import Arm.Memory.Attr
+import Lean
+import Lean.Meta.Tactic.Rewrite
+import Lean.Meta.Tactic.Rewrites
+import Lean.Elab.Tactic.Conv
+import Lean.Elab.Tactic.Conv.Basic
+
+open Lean Meta Elab Tactic
+
+
+/-! ## Memory Separation Automation
+
+##### A Note on Notation
+
+- `[a..an)`: a range of memory starting with base address `a` of length `an`.
+  aka `mem_legal a an`.
+- `[a..an) ⟂ [b..bn)`: `mem_disjoint a an b bn`.
+- `[a..an] ⊆ [b..bn)`: `mem_subset a an b bn`
+
+##### Tactic Loop
+
+The core tactic tries to simplify expressions of the form:
+`mem.write_bytes [b..bn) val |>. read_bytes [a..an)`
+by case splitting:
+
+1. If `[a..an) ⟂ [b..bn)`, the write does not alias the read,
+  and can be replaced with ` mem.read_bytes [a..an) `
+2. If `[a..an] ⊆ [b..bn)`, the write aliases the read, and can be replaced with
+  `val.extractLsBs adjust([a..an), [b..bn))`. Here, `adjust` is a function that
+  adjusts the read indices `[a..an)` with respect to the write indices `[b..bn)`,
+  to convert a read from `mem` into a read from `val`.
+
+The tactic shall be implemented as follows:
+1. Search the goal state for `mem.write_bytes [b..bn) val |>.read_bytes [a/..an)`.
+2. Try to prove that either `[a..an) ⟂ [b..bn)`, or `[a..an) ⊆ [b..bn)`.
+    2a. First search the local context for assumptions of this type.
```

Comment:
Didn't read code yet; you may have already done this but just in case:

If no mem separate/subset assumptions are present, then throw an error to tell the user that we expect them to specify such assumptions for all memory regions of interest.  LNSym doesn't support automated verification of programs that do dynamic memory allocation.

Reply:
Also, TODO for later: if any non-primed separate/subset assumptions are detected, error out to tell the user that no automation is supported in this case.

Reply:
Added these to the document explaining the vision of the tactic.